framework/include/utils/MooseUtils.h

```diff
@@ -1197,6 +1198,96 @@ isDigits(const std::string & str)
 {
   return std::all_of(str.begin(), str.end(), [](unsigned char c) { return std::isdigit(c); });
 }
+
+/**
+ * This function should not be called directly (although it can be),
+ * instead see the moose::map_find() macro.
+ *
+ * Calls find(key), and checks the result against end(). Returns the
+ * corresponding value if found, throws an error otherwise. Templated
+ * on the type of map, so this will work with both std::map and
+ * std::unordered_map.
+ * Note:
+ * - we need both a const and non-const version for the return type, due to the use
+ *   of the find operation to simply find a value or modify a value
+ * - const references can bind to more things for templated function parameter Key
+ * - we cannot have the exact same signatures and different return types
+ * - typename T::type requests that T defines the type function
+ */
+template <
+    typename Map,
+    typename Key,
+    typename T,
+    typename T::type,
+    typename T::name,
+    typename std::enable_if<!libMesh::Utility::is_streamable<Key>::value, Key>::type * = nullptr>
+inline typename Map::mapped_type &
+map_find(Map & map, const Key & key, const T * object, const char * filename, int line_number)
+{
+  auto it = map.find(key);
+  if (it == map.end())
+    mooseError("map_find() error: key not found during a search in class ",
+               object->name(),
+               " of type ",
+               object->type(),
+               ".\n See file ",
+               filename,
+               " at line ",
+               line_number);
+  return it->second;
+}
+template <
+    typename Map,
+    typename Key,
+    typename T,
+    typename T::type,
+    typename T::name,
+    typename std::enable_if<!libMesh::Utility::is_streamable<Key>::value, Key>::type * = nullptr>
+inline const typename Map::mapped_type &
+map_find(const Map & map, const Key & key, const T * object, const char * filename, int line_number)
+{
+  auto it = map.find(key);
+  if (it == map.end())
+    mooseError("map_find() error: key not found during a search in class ",
+               object->name(),
+               " of type ",
+               object->type(),
+               ".\n See file ",
+               filename,
+               " at line ",
+               line_number);
+  return it->second;
+}
+
+/**
+ * Version where this is being used outside of a class so `this` does not exist
+ * First template for non-const returns
+ * Second template for const returns
+ * We enforce that the const-ness of the Map matches the constness of the return argument
+ */
```

Comment:
I don't get this. Who calls this?

Reply:
this is called by the moose_map_find macro if the above substitution failed

Reply:
ok, after your push it makes lightly more sense, but you have the enable ifs backwards I think

Reply:
Nonetheless, if `this` does not exist the macro substitution will give you a non-existing symbol right away.

Reply:
yeah. Not much I can do about this?